### PR TITLE
C6C2 whole-project PR for review

### PR DESCRIPTION
**Pull request by mariatens**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,7 +1,87 @@
...
+import { useState, useEffect } from "react";
+import { IUser } from "./interfaces";
+import { Routes, Route, NavLink } from "react-router-dom";
+import { IResource } from "./interfaces";
+import axios from "axios";
+import { Resource } from "./Pages/Resource";
+import "./App.css";
+import { useLocalStorage } from "./utils/localStorage";
+
+export const url =
```
Comment (GracieLouFreebush22) [inline src/App.tsx:13]: Tip: be careful about using 'url' as a variable name here, maybe better with "repoUrl", "baseUrl" etc to limit confusion with build in JS name for URL

File: src/components/ResourcePreview.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    await axios.post(likeURL, { like: updatedDisLikeStatus, userId: userid });
+
+    setIsLiked(updatedDisLikeStatus);
+  };
+
+  return (
+    <div className="resourcePreview">
+      <h3 className="resourceTitle"> {props.resource.resource_name} </h3>
+      <Link className="link" to={`/resource/${props.resource.resource_id}`}>
+        <button className="button-30"> Show more! </button>
```
Comment (GracieLouFreebush22) [inline src/components/ResourcePreview.tsx:56]: naming of the button class name is vague

File: src/components/ResourcePreview.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import { AiFillDislike } from "react-icons/ai";
+import { AiOutlineLike } from "react-icons/ai";
+import { AiOutlineDislike } from "react-icons/ai";
+
+interface IResourcePreview {
+  resource: IResource;
+  userID: number | null;
+}
+
+export default function ResourcePreview(props: IResourcePreview): JSX.Element {
```
Comment (GracieLouFreebush22) [inline src/components/ResourcePreview.tsx:16]: Possible to decrease word count for rest of file if props: IResourcePreview had been destructured (props.resource.thing) could've been simplified throughout

File: src/components/NewComment.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { url } from "../App";
+
+interface INewCommentProps {
+  userID: number;
+  resourceID: number;
+  submitted: React.Dispatch<React.SetStateAction<boolean>>;
+}
+
+export default function NewComment(props: INewCommentProps): JSX.Element {
+  const userID = props.userID;
```
Comment (GracieLouFreebush22) [inline src/components/NewComment.tsx:12]: can also destructure both of these on one line, could be simpler to read (it's still easy to read :)

File: src/components/NewComment.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  const resourceID = props.resourceID;
+
+  const [commentText, setCommentText] = useState<string>("");
+
+  async function handleSubmit() {
+    props.submitted(false);
+    if (!commentText) {
+      window.alert("No comment text!");
+    } else {
+      await axios.post(`${url}/comments/${resourceID}/`, {
```
Comment (GracieLouFreebush22) [inline src/components/NewComment.tsx:22]: on the resource preview page the axios url was contained in part in a variable, could be good to keep this consistent across component files.

File: src/components/SearchBar.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
+interface SearchBarProps {
```
Comment (GracieLouFreebush22) [inline src/components/SearchBar.tsx:1]: well done for keeping this component simple and straightforward

File: src/App.tsx [hunk 7]
```diff
@@ -1,7 +1,87 @@
...
+    ? "https://coding-resources-backend.onrender.com"
+    : "http://localhost:4000";
 
 function App(): JSX.Element {
-  return <h1>{greet("World")}</h1>;
+  const [userID, setUserID] = useLocalStorage("userID", null);
+  const [users, setUsers] = useState<IUser[]>([]);
+  const [allResources, setAllResources] = useState<IResource[]>([]);
+  useEffect(() => {
+    const userNamesCompleteURL = url + "/users";
```
Comment (GracieLouFreebush22) [inline src/App.tsx:23]: like the use renaming here, made this easy to read

File: src/App.tsx [hunk 8]
```diff
@@ -1,7 +1,87 @@
...
+      const { data } = await axios.get(userNamesCompleteURL);
+      setUsers(data);
+    };
+    fetchUserNames();
+    fetchAllResources();
+  }, []);
+
+  return (
+    <div>
+      <div className="navbar">
```
Comment (GracieLouFreebush22) [inline src/App.tsx:40]: nav bar could have been it's own component to reduce length on this file and increase modularity

File: src/Pages/Home.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
+import { useState } from "react";
+import ResourcePreview from "../components/ResourcePreview";
+import SearchBar from "../components/SearchBar";
+import TagFilter from "../components/TagFilter";
+import { IUser } from "../interfaces";
```
Comment (GracieLouFreebush22) [inline src/Pages/Home.tsx:5]: lines 5 and 6 could have been combined

File: src/Pages/Home.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  users,
+  userID,
+}: IHomePageProps): JSX.Element {
+  const [searchText, setSearchText] = useState<string>("");
+  const [selectedTags, setSelectedTags] = useState<string[]>([]);
+
+  const handleFilterTag = (clickedTag: string) => {
+    const currentTags = selectedTags;
+
+    if (!currentTags.includes(clickedTag)) {
```
Comment (GracieLouFreebush22) [inline src/Pages/Home.tsx:28]: good use of different methods here <3

File: src/Pages/Home.tsx [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+      ];
+    }
+    setSelectedTags(
+      currentTags.filter((tag) => {
+        return tag;
+      })
+    );
+  };
+
+  const [switchFromOrToAnd, setSwitchFromOrToAnd] = useState<boolean>(false);
```
Comment (GracieLouFreebush22) [inline src/Pages/Home.tsx:44]: try to keep all pieces of state in one area, makes it easier to read

File: src/Pages/Resource.tsx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  const filteredUser = users.filter(
+    (user) => user.user_id === oneResource.user_id
+  );
+
+  const handleAddToDoList = async (userID: number, resourceID: number) => {
+    await axios.post(url + "/to-do-list", {
+      resource_id: resourceID,
+      user_id: userID,
+    });
+    window.alert("Added the post to your to do list");
```
Comment (GracieLouFreebush22) [inline src/Pages/Resource.tsx:56]: loved the use of the window alerts!

File: src/Pages/NewResource.tsx [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+      tags: tagsArray.filter((tag) => {
+        return tag;
+      }),
+    });
+  }
+
+  return (
+    <div>
+      <h1>Create a new resource</h1>
+      <form
```
Comment (GracieLouFreebush22) [inline src/Pages/NewResource.tsx:91]: tip: you can use the 'required' attribute on form inputs instead of using conditional rendering to prevent empty fields on form submissions.

File: src/utils/formatSubmissionDate.test.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import formatSubmissionDate from "./formatSubmissionDate";
+
+test("function removes unnecessary time from end of date", () => {
+  expect(formatSubmissionDate("2023-01-17T00:00:00.000Z")).toEqual(
```
Comment (GracieLouFreebush22) [inline src/utils/formatSubmissionDate.test.ts:4]: to beef this up can always consider edge cases :)

File: src/utils/formatTags.ts [hunk 15]
```diff
@@ -0,0 +1,11 @@
+export function formatTags(tags: string[]): string[] {
+  const actualArray = tags[0];
```
Comment (GracieLouFreebush22) [inline src/utils/formatTags.ts:2]: full or both parts array might be better name? Not sure as to why the threshold is two? think naming could be slightly improved here

File: src/utils/formatTags.test.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import { formatTags } from "./formatTags";
+
+const tags = [
+  '{"React","Typescript","Javascript","Front-end","Back-end","CSS","HTML","SQL"}',
+];
+
+const empty = ["{}"];
```
Comment (GracieLouFreebush22) [inline src/utils/formatTags.test.ts:7]: emptyTags or emptyObj , needs a friend on the end of the variable name to clarify

File: src/utils/localStorage.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import { useState, useEffect } from "react";
+
+function getStorageValue(key: string, defaultValue: unknown) {
```
Comment (GracieLouFreebush22) [inline src/utils/localStorage.ts:3]: cool use of unknown, hadn't seen this before

File: src/utils/localStorage.ts [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  // getting stored value
+  const saved = localStorage.getItem(key);
+  if (typeof saved === "string") {
```
Comment (GracieLouFreebush22) [inline src/utils/localStorage.ts:6]: the use of 'saved', 'initial' and 'key' here could have more specific names to help the reader better understand what localStorage is specifically being used for. We can see its for items, but what items (resource, tag, type etc)?

File: src/index.tsx [hunk 19]
```diff
@@ -1,13 +1,16 @@
 import React from "react";
 import ReactDOM from "react-dom/client";
 import App from "./App";
+import { BrowserRouter } from "react-router-dom";
 
 const root = ReactDOM.createRoot(
   document.getElementById("root") as HTMLElement
 );
```
Comment (GracieLouFreebush22) [inline src/index.tsx:8]: so fresh n so clean WOWEEEEE

**Review by GracieLouFreebush22 — COMMENTED** [review]
nice structure overall, it was very clear to see the component tree vision. We found the code really easy to read, and generally bit of like-minded code stuck together which helped. Also enjoyed seeing widespread use of ternary expressions, abstracting where it is needed, and prioritising modularity. Could have been extra helpful to see some signposting comments on the general purpose of each section (specifically in the bigger page files). Liked the use of separating out the page vs component files as well. Key themes in our review was 1) keeping consistent with props 2) improving naming of variables. WELL DONE Y'ALL. YEEHAW!
